### ISSUE TITLE
Pin environments used in CI

### ISSUE THREAD

**Issue by hdgarrood** [purescript/purescript#4107]
Right now, we use `ubuntu-latest`, `macos-latest` etc in CI, in `.github/workflows/ci.yaml`. This is problematic, because it means that the prebuilt binaries are likely to stop working on older platforms between releases, without anything in the changelog. See for example https://discourse.purescript.org/t/dockerfile-for-purescript-0-14-2/2379, in which the prebuilt binaries no longer worked because we are linking against a newer version of glibc.

I think we should probably pin these to specific versions, and generally be a bit more conservative about updating, since binaries built on a newer OS are sometimes unlikely to work on an older one.

**Comment by rhendric** [comment]
FWIW, the pre-0.14.2 Travis config built on Ubuntu Trusty, macOS 10.13, and Windows Server version 1809. None of these OS versions are exactly [available in GitHub Actions](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources): the earliest Ubuntu available is Bionic (`ubuntu-18.04`), the earliest macOS is 10.15 (`macos-10.15`, which is currently also `macos-latest`), and the Windows offerings are all long-term rather than SAC (we can choose `windows-2016` or `windows-2019` = `windows-latest`).

So if anyone was wondering, as I was, if we could just pin the GitHub versions to what they were before and push out a quick 0.14.3 to fix this... well, not really. If we're lucky, Bionic might still be compatible with whatever glibc is in everyone's Docker images, but someone'd have to check.

(Actually, it looks like `ubuntu-16.04` is still nominally available until [September 20](https://github.blog/changelog/2021-04-29-github-actions-ubuntu-16-04-lts-virtual-environment-will-be-removed-on-september-20-2021/), so it's technically an option to do that but we'd basically be committing to a major release between now and then, assuming our new policy for bumping CI versions is ‘only on major releases’.)

Let's also note that GitHub's window for decommissioning runner OSes looks narrower than Travis's was. That's one reason—aside from plain old ‘what could go wrong?’ ignorance—I used `-latest` versions in the configs; repinning versions seems like it would be a somewhat frequently-revisited chore performed in a scramble when things break. But if we add it to the major release procedure, probably not so bad.

**Comment by hdgarrood** [comment]
I think once the Ubuntu version in question is EOL (as ubuntu 16.04 now is), it's fine to drop support for it, so I think having the Ubuntu job run on 18.04 sounds good. People can still build the compiler themselves if they are not ready to move off EOL'd distributions.

**Comment by hdgarrood** [comment]
The question of when we are allowed to bump these things is a good one, though. I would be fine with noting it in the changelog when bumping these versions but not treating it as a breaking change, because it only affects the prebuilt binaries, not the compiler itself - again, because people can always build the compiler themselves if they need to.

**Comment by FloWi** [comment]
> I think once the Ubuntu version in question is EOL (as ubuntu 16.04 now is), it's fine to drop support for it, so I think having the Ubuntu job run on 18.04 sounds good. People can still build the compiler themselves if they are not ready to move off EOL'd distributions.

A build on ubuntu 18.04 sounds fine to me since it uses the same glibc version as the official node docker-images. Imho it's valuable to be able to run the purescript toolchain on the official node-images, wdyt?

**Comment by hdgarrood** [comment]
Yes, I agree that being able to run on the official node images is valuable.

**Comment by rhendric** [comment]
But, to be clear, not so valuable that we aren't going to make promises about it?

There's no guarantee that, in the future, GitHub won't drop Linux runner OSes before the Node Docker images do. If we cared about maintaining that property, we should be running our Linux builds inside our own Docker image or something.

(I don't know what we'd do as an equivalent for macOS and Windows, but also, do they have stronger backwards-compatibility guarantees than Linux does? If so, maybe we can leave the GitHub runners pinned at `latest` so that they don't have to be bumped, and use a Docker image to fix the Ubuntu version we build against which we bump on whatever schedule we desire.)

**Comment by hdgarrood** [comment]
Yeah, so I'm suggesting the following strategy:
- Use specific OS versions for all CI runners
- Update to newer versions when at least one of the following is true:
  - We are forced to (because GitHub drops support for them)
  - The version that is currently in use is EOL'd
  - There is some benefit from upgrading that we deem worth the breakage, if any.

There is of course no guarantee that we will end up with binaries which run on the Node Docker images this way. I think even downgrading to 18.04 might not be sufficient to get the prebuilt binaries running in all of the official Node images, because it appears that node:14 and node:12 use Debian 9 "stretch", which has glibc 2.24, whereas Ubuntu 18.04 has glibc 2.27. But on the other hand, it's still easy to compile your own compiler if the prebuilt binaries don't work for you, and I would prefer not to increase our own maintenance burden more than necessary by trying to support lots of configurations.

**Comment by fabrizio-cn** [comment]
I don't think it's worth complicating the build process by building in Docker (we did it for years for Spago on Travis, and it was not exactly pleasant) just to chase the compatibility with the Node image. At that point we might want to just try to go for a static linux build?

In the meanwhile, I think it would be valuable to release a `0.14.3` with a downgraded `glibc` as soon as we can, so I opened #4113

**Comment by f-f** [comment]
I don't think it's worth complicating the build process by building in Docker (we did it for years for Spago on Travis, and it was not exactly pleasant) just to chase the compatibility with the Node image. At that point we might want to just try to go for a static linux build?

In the meanwhile, I think it would be valuable to release a `0.14.3` with a downgraded `glibc` as soon as we can, so I opened #4113 to pin the OS versions in CI.
